Unfortunately `./python.exe -m test -v test_opcache -m test_binary_op -R 3:3` is failing:

<details>

```python
== CPython 3.14.0a4+ (heads/fix-zero-division-dirty:850a862a7d8, Jan 18 2025, 13:17:14) [Clang 16.0.0 (clang-1600.0.26.4)]
== macOS-15.1.1-arm64-arm-64bit-Mach-O little-endian
== Python build: debug
== cwd: /Users/eclips4/programming/programming-languages/cpython/build/test_python_worker_4826æ
== CPU count: 8
== encodings: locale=UTF-8 FS=utf-8
== resources: all test resources are disabled, use -u option to unskip tests

Using random seed: 548978877
0:00:00 load avg: 2.19 Run 1 test sequentially in a single process
0:00:00 load avg: 2.19 [1/1] test_opcache
beginning 6 repetitions. Showing number of leaks (. for 0 or less, X for 10 or more)
123:456
test_binary_op (test.test_opcache.TestSpecializer.test_binary_op) ... ok

----------------------------------------------------------------------
Ran 1 test in 0.006s

OK
Xtest_binary_op (test.test_opcache.TestSpecializer.test_binary_op) ... ok

----------------------------------------------------------------------
Ran 1 test in 0.007s

OK
Xtest_binary_op (test.test_opcache.TestSpecializer.test_binary_op) ... FAIL

======================================================================
FAIL: test_binary_op (test.test_opcache.TestSpecializer.test_binary_op)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/eclips4/programming/programming-languages/cpython/Lib/test/test_opcache.py", line 1405, in test_binary_op
    binary_op_nan()
    ~~~~~~~~~~~~~^^
  File "/Users/eclips4/programming/programming-languages/cpython/Lib/test/test_opcache.py", line 1402, in binary_op_nan
    self.assert_no_opcode(compactlong_lhs, "BINARY_OP_EXTEND")
    ~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/eclips4/programming/programming-languages/cpython/Lib/test/test_opcache.py", line 42, in assert_no_opcode
    self.assertNotIn(opname, opnames)
    ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^
AssertionError: 'BINARY_OP_EXTEND' unexpectedly found in {'RESUME_CHECK', 'RETURN_VALUE', 'POP_TOP', 'LOAD_CONST_IMMORTAL', 'BINARY_OP_EXTEND', 'LOAD_SMALL_INT', 'LOAD_FAST'}

----------------------------------------------------------------------
Ran 1 test in 0.012s

FAILED (failures=1)
test test_opcache failed
test_opcache failed (1 failure)

== Tests result: FAILURE ==

1 test failed:
    test_opcache

Total duration: 174 ms
Total tests: run=1 (filtered) failures=1
Total test files: run=1/1 (filtered) failed=1
Result: FAILURE
```
</details>

I'm digging into what's happening.

--- a/Lib/test/test_opcache.py
+++ b/Lib/test/test_opcache.py
@@ -1395,9 +1395,11 @@ class TestSpecializer(TestBase):
                 arg / 42
 
             compactlong_lhs(1.0)
-            compactlong_lhs(float('nan'))
+            for _ in range(100):
+                compactlong_lhs(float('nan'))
             compactlong_rhs(1.0)
-            compactlong_rhs(float('nan'))
+            for _ in range(100):
+                compactlong_rhs(float('nan'))
 
             self.assert_no_opcode(compactlong_lhs, "BINARY_OP_EXTEND")
             self.assert_no_opcode(compactlong_rhs, "BINARY_OP_EXTEND")

--- a/Lib/test/test_opcache.py
+++ b/Lib/test/test_opcache.py
@@ -1384,22 +1384,26 @@ class TestSpecializer(TestBase):
 
         def binary_op_nan():
             def compactlong_lhs(arg):
-                42 + arg
-                42 - arg
-                42 * arg
-                42 / arg
+                return (
+                    42 + arg,
+                    42 - arg,
+                    42 * arg,
+                    42 / arg,
+                )
             def compactlong_rhs(arg):
-                arg + 42
-                arg - 42
-                arg * 42
-                arg / 42
-
-            compactlong_lhs(1.0)
+                return (
+                    arg + 42,
+                    arg - 42,
+                    arg * 2,
+                    arg / 42,
+                )
+            nan = float('nan')
+            self.assertEqual(compactlong_lhs(1.0), (43.0, 41.0, 42.0, 42.0))
             for _ in range(100):
-                compactlong_lhs(float('nan'))
-            compactlong_rhs(1.0)
+                self.assertTrue(all(filter(lambda x: x is nan, compactlong_lhs(nan))))
+            self.assertEqual(compactlong_rhs(42.0), (84.0, 0.0, 84.0, 1.0))
             for _ in range(100):
-                compactlong_rhs(float('nan'))
+                self.assertTrue(all(filter(lambda x: x is nan, compactlong_rhs(nan))))
 
             self.assert_no_opcode(compactlong_lhs, "BINARY_OP_EXTEND")
             self.assert_no_opcode(compactlong_rhs, "BINARY_OP_EXTEND")

--- a/Python/specialize.c
+++ b/Python/specialize.c
@@ -2421,14 +2421,14 @@ float_compactlong_guard(PyObject *lhs, PyObject *rhs)
 {
     return (
         PyFloat_CheckExact(lhs) &&
+        !isnan(PyFloat_AsDouble(lhs)) &&
         PyLong_CheckExact(rhs) &&
-        _PyLong_IsCompact((PyLongObject *)rhs) &&
-        !isnan(PyFloat_AsDouble(lhs))
+        _PyLong_IsCompact((PyLongObject *)rhs)
     );
 }
 
 static inline int
-float_compactlong_guard_true_div(PyObject *lhs, PyObject *rhs)
+nonzero_float_compactlong_guard(PyObject *lhs, PyObject *rhs)
 {
     return (
         float_compactlong_guard(lhs, rhs) && !PyLong_IsZero(rhs)
@@ -2455,15 +2455,15 @@ static inline int
 compactlong_float_guard(PyObject *lhs, PyObject *rhs)
 {
     return (
-        PyFloat_CheckExact(rhs) &&
         PyLong_CheckExact(lhs) &&
         _PyLong_IsCompact((PyLongObject *)lhs) &&
+        PyFloat_CheckExact(rhs) &&
         !isnan(PyFloat_AsDouble(rhs))
     );
 }
 
 static inline int
-compactlong_float_guard_true_div(PyObject *lhs, PyObject *rhs)
+nonzero_compactlong_float_guard(PyObject *lhs, PyObject *rhs)
 {
     return (
         compactlong_float_guard(lhs, rhs) && PyFloat_AsDouble(rhs) != 0.0
@@ -2487,14 +2487,14 @@ LONG_FLOAT_ACTION(compactlong_float_true_div, /)
 static _PyBinaryOpSpecializationDescr float_compactlong_specs[NB_OPARG_LAST+1] = {
     [NB_ADD] = {float_compactlong_guard, float_compactlong_add},
     [NB_SUBTRACT] = {float_compactlong_guard, float_compactlong_subtract},
-    [NB_TRUE_DIVIDE] = {float_compactlong_guard_true_div, float_compactlong_true_div},
+    [NB_TRUE_DIVIDE] = {nonzero_float_compactlong_guard, float_compactlong_true_div},
     [NB_MULTIPLY] = {float_compactlong_guard, float_compactlong_multiply},
 };
 
 static _PyBinaryOpSpecializationDescr compactlong_float_specs[NB_OPARG_LAST+1] = {
     [NB_ADD] = {compactlong_float_guard, compactlong_float_add},
     [NB_SUBTRACT] = {compactlong_float_guard, compactlong_float_subtract},
-    [NB_TRUE_DIVIDE] = {compactlong_float_guard_true_div, compactlong_float_true_div},
+    [NB_TRUE_DIVIDE] = {nonzero_compactlong_float_guard, compactlong_float_true_div},
     [NB_MULTIPLY] = {compactlong_float_guard, compactlong_float_multiply},
 };
 

--- a/Python/specialize.c
+++ b/Python/specialize.c
@@ -2416,8 +2416,6 @@ binary_op_fail_kind(int oparg, PyObject *lhs, PyObject *rhs)
 
 /* float-long */
 
-// Guards should check that the float part is not NaN.
-// Guards for / (aka true_div) should check for 0 or 0.0 as the rhs.
 static inline int
 float_compactlong_guard(PyObject *lhs, PyObject *rhs)
 {

